### PR TITLE
Adding Base structure for pod/container config

### PR DESCRIPTION
**Pull request by CharlyF**
Goal is to have

```yaml
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  credentials:
    apiKey: <DATADOG_API_KEY>
    appKey: <DATADOG_APP_KEY>
  features:
    orchestratorExplorer:
      enabled: true
    kubeStateMetricsCore:
      enabled: true
    apm:
      enabled: true
    dogstastd:
      enabled: true
  global: # global settings 
    clusterName: foo
    registry: gcr.io/datadoghq
    provider:
      gke: 
        autopilot: true
  override:
    nodeAgent:
      podOverride: 
        labels:
          provider: datadog
        image: gcr.io/datadoghq/agent:latest-jmx
        containers:
            name: trace-agent: 
            resources:
              requests:
                cpu: 2
            env:
            - name: "DD_DEBUG_LEVEL"
              value: "TRACE"
    clusterAgent: # override for the agent deployment pod
      podOverride: 
        containers: []
[...]
    clusterCheckRunner:
[...]
```

### What does this PR do?

Base to build upon to release the V2 of the operator.

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:87]: todo: update the comment

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:106]: todo: update the comment

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:109]: todo: update the comment

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:120]: Todo: update comment + remove link to the doc

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:128]:
```suggestion
	// Specify additional volume mounts in the container.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:139]:
```suggestion
	// Command allows the specification of custom entrypoint for the container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:152]:
```suggestion
	// Configure the Readiness Probe of the container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:156]:
```suggestion
	// Configure the Liveness Probe of the container
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml
Comment (clamoriniere) [inline config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:9]: this part should have been added back

**Review by clamoriniere — COMMENTED** [review]
Looks good it term of structure definition. The fields descriptions need to be updated

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:74]: Should we name it `DatadogAgentPodTemplateOverride` ?

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:65]: Should we rename it as `DatadogAgentResourceOverride`?

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (clamoriniere) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:35]:
```suggestion
	// Override the configuration of the components
```

File: docs/configuration.md
Comment (CharlyF) [reply docs/configuration.md]: that's a good point - I will tackle right after we merge this if that's OK

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:17]:
```suggestion
	// NodeAgentResourceName is the name of the Datadog Node Agent
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:22]:
```suggestion
	ClusterChecksRunnerResourceName ResourceName = "clusterChecksRunner"
```

(to be consistent with what we do in Helm https://github.com/DataDog/helm-charts/blob/main/charts/datadog/values.yaml#L1169)

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:27]:
```suggestion
	// Features running on the Agent and Cluster Agent
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:31]:
```suggestion
	// Global settings to configure the agents
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:35]:
```suggestion
	// Override the default configurations of the agents
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:66]:
```suggestion
	// DatadogAgentPodTemplateOverride is used to configure the components at the pod level in an agnostic way
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:75]:
```suggestion
	// Configure the basic configurations for each agent container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:87]:
```suggestion
	// Configure the different component (Datadog Agent, Cluster Agent, Cluster Check Runner) tolerations.
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml
Comment (clamoriniere) [reply config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:9]: ok,
maybe you could add them (duplicated from the v1alpha1)

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:106]:
```suggestion
	// Annotations provide annotations that will be added to the different component (Datadog Agent, Cluster Agent, Cluster Check Runner) pods.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:109]:
```suggestion
	// AdditionalLabels provide labels that will be added to the different component (Datadog Agent, Cluster Agent, Cluster Check Runner) pods.
```

**Review by clamoriniere — APPROVED** [review]
looks good on my side, when others comments will be fixed

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:117]:
```suggestion
	// Name of the container that is overriden
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:121]: Specify additional environmental variables in the container

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:136]: should we write something more generic here? (or, consider having this on a second line and write something generic above it)

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:140]:
```suggestion
	// Command allows the specification of a custom entrypoint for container
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:144]:
```suggestion
	// Args allows the specification of extra args to the `Command` parameter
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:148]:
```suggestion
	// HealthPort of the container for the internal liveness probe.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:149]:
```suggestion
	// Must be the same as the Liveness/Readiness probes.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:167]:
```suggestion
	// Use "datadog/dogstatsd:latest" for standalone Datadog Agent DogStatsD6
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:173]:
```suggestion
	// Define the image tag to use.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:180]: should we use `JMXEnabled`? seems most places we capitalize 'JMX'

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:186]:
```suggestion
	// It is possible to specify Docker registry credentials.
```

File: apis/datadoghq/v2alpha1/datadogagent_types.go
Comment (celenechang) [inline apis/datadoghq/v2alpha1/datadogagent_types.go:201]:
```suggestion
// DatadogAgent Deployment with the Datadog Operator.
```

**Comment by celenechang** [comment]
Thanks for doing this. I noticed that for the `nodeAgent`, there is a `podOverride` level above `containers`, but for the `clusterAgent`, `containers` is the next level down. Is this something we could make consistent?

```
   nodeAgent:
      podOverride: 
        containers:
            ...
    clusterAgent:
      containers:
```

**Comment by clamoriniere** [comment]
> Thanks for doing this. I noticed that for the `nodeAgent`, there is a `podOverride` level above `containers`, but for the `clusterAgent`, `containers` is the next level down. Is this something we could make consistent?
> 
> ```
>     nodeAgent:
>       podOverride: 
>         containers:
>             ...
>     clusterAgent:
>       containers:
> ```

Yep the example is not valide. the structure inside `clusterAgent` is  the same than for `nodeAgent`. we can update the PR description

**Review by celenechang — APPROVED** [review]

**Pull request merged**
